File: src/Web/Stocker.Web.Admin/Services/SubscriptionService.cs
namespace Stocker.Web.Admin.Services;

public interface ISubscriptionService
{
    Task<SubscriptionsResponse?> GetSubscriptionsAsync(Guid? tenantId = null, string? status = null, bool? autoRenew = null);
    Task<SubscriptionDto?> GetSubscriptionByIdAsync(Guid id);
    Task<SubscriptionDto?> CreateSubscriptionAsync(CreateSubscriptionRequest request);
    Task<bool> UpdateSubscriptionAsync(Guid id, UpdateSubscriptionRequest request);
    Task<bool> CancelSubscriptionAsync(Guid id, string reason);
}

public class SubscriptionService : ISubscriptionService
{
    private readonly IApiService _apiService;
    private readonly ILogger<SubscriptionService> _logger;

    public SubscriptionService(IApiService apiService, ILogger<SubscriptionService> logger)
    {
        _apiService = apiService;
        _logger = logger;
    }

    public async Task<SubscriptionsResponse?> GetSubscriptionsAsync(Guid? tenantId = null, string? status = null, bool? autoRenew = null)
    {
        try
        {
            var queryParams = new List<string>();
            if (tenantId.HasValue)
                queryParams.Add($"tenantId={tenantId}");
            if (!string.IsNullOrEmpty(status))
                queryParams.Add($"status={status}");
            if (autoRenew.HasValue)
                queryParams.Add($"autoRenew={autoRenew}");

            var query = queryParams.Any() ? "?" + string.Join("&", queryParams) : "";
            
            var response = await _apiService.GetAsync<ApiResponse<List<SubscriptionDto>>>($"/api/master/subscriptions{query}");
            if (response?.Success == true && response.Data != null)
            {
                return new SubscriptionsResponse
                {
                    Subscriptions = response.Data,
                    TotalCount = response.Data.Count
                };
            }
            return null;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting subscriptions");
            return null;
        }
    }

    public async Task<SubscriptionDto?> GetSubscriptionByIdAsync(Guid id)
    {
        try
        {
            var response = await _apiService.GetAsync<ApiResponse<SubscriptionDto>>($"/api/master/subscriptions/{id}");
            return response?.Success == true ? response.Data : null;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting subscription {SubscriptionId}", id);
            return null;
        }
    }

    public async Task<SubscriptionDto?> CreateSubscriptionAsync(CreateSubscriptionRequest request)
    {
        try
        {
            var response = await _apiService.PostAsync<ApiResponse<SubscriptionDto>>("/api/master/subscriptions", request);
            return response?.Success == true ? response.Data : null;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating subscription");
            return null;
        }
    }

    public async Task<bool> UpdateSubscriptionAsync(Guid id, UpdateSubscriptionRequest request)
    {
        try
        {
            var response = await _apiService.PutAsync<ApiResponse<bool>>($"/api/master/subscriptions/{id}", request);
            return response?.Success == true && response.Data;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating subscription {SubscriptionId}", id);
            return false;
        }
    }

    public async Task<bool> CancelSubscriptionAsync(Guid id, string reason)
    {
        try
        {
            var request = new { reason };
            var response = await _apiService.PostAsync<ApiResponse<bool>>($"/api/master/subscriptions/{id}/cancel", request);
            return response?.Success == true && response.Data;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error cancelling subscription {SubscriptionId}", id);
            return false;
        }
    }
}

// DTOs
public class SubscriptionsResponse
{
    public List<SubscriptionDto> Subscriptions { get; set; } = new();
    public int TotalCount { get; set; }
}

public class SubscriptionDto
{
    public Guid Id { get; set; }
    public Guid TenantId { get; set; }
    public Guid PackageId { get; set; }
    public string SubscriptionNumber { get; set; } = string.Empty;
    public string TenantName { get; set; } = string.Empty;
    public string PackageName { get; set; } = string.Empty;
    public string Status { get; set; } = string.Empty;
    public string BillingCycle { get; set; } = string.Empty;
    public decimal Price { get; set; }
    public string Currency { get; set; } = "TRY";
    public DateTime StartDate { get; set; }
    public DateTime CurrentPeriodStart { get; set; }
    public DateTime CurrentPeriodEnd { get; set; }
    public DateTime? TrialEndDate { get; set; }
    public DateTime? CancelledAt { get; set; }
    public string? CancellationReason { get; set; }
    public bool AutoRenew { get; set; }
    public int UserCount { get; set; }
    public List<SubscriptionModuleDto> Modules { get; set; } = new();
}

public class SubscriptionModuleDto
{
    public Guid Id { get; set; }
    public string ModuleCode { get; set; } = string.Empty;
    public string ModuleName { get; set; } = string.Empty;
    public int? MaxEntities { get; set; }
    public bool IsActive { get; set; }
}

public class CreateSubscriptionRequest
{
    public Guid TenantId { get; set; }
    public Guid PackageId { get; set; }
    public int BillingCycle { get; set; } = 0; // 0 = Monthly
    public decimal? CustomPrice { get; set; }
    public string Currency { get; set; } = "TRY";
    public DateTime? StartDate { get; set; }
    public int? TrialDays { get; set; }
    public bool AutoRenew { get; set; } = true;
    public int UserCount { get; set; } = 1;
}

public class UpdateSubscriptionRequest
{
    public int? BillingCycle { get; set; }
    public decimal? Price { get; set; }
    public string? Currency { get; set; }
    public bool? AutoRenew { get; set; }
    public int? UserCount { get; set; }
}

File: stocker-admin/src/services/api/subscriptionService.ts
import { apiClient } from './apiClient';

// Subscription DTOs matching backend
export interface SubscriptionDto {
  id: string;
  tenantId: string;
  packageId: string;
  subscriptionNumber: string;
  tenantName: string;
  packageName: string;
  status: SubscriptionStatus;
  billingCycle: BillingCycle;
  price: number;
  currency: string;
  startDate: string;
  currentPeriodStart: string;
  currentPeriodEnd: string;
  trialEndDate?: string;
  cancelledAt?: string;
  cancellationReason?: string;
  autoRenew: boolean;
  userCount: number;
  modules: SubscriptionModuleDto[];
  usages: SubscriptionUsageDto[];
}

export interface SubscriptionModuleDto {
  id: string;
  subscriptionId: string;
  moduleCode: string;
  moduleName: string;
  maxEntities?: number;
  isActive: boolean;
}

export interface SubscriptionUsageDto {
  id: string;
  subscriptionId: string;
  metricName: string;
  value: number;
  recordedAt: string;
}

export enum SubscriptionStatus {
  Deneme = 0,
  Aktif = 1,
  Askida = 2,
  Iptal = 3,
  Suresi_Dolmus = 4
}

export enum BillingCycle {
  Aylik = 0,
  Yillik = 1,
  Tek_Sefer = 2
}

export interface GetSubscriptionsQuery {
  tenantId?: string;
  status?: SubscriptionStatus;
  autoRenew?: boolean;
}

export interface CreateSubscriptionCommand {
  tenantId: string;
  packageId: string;
  billingCycle: BillingCycle;
  startDate: string;
  autoRenew?: boolean;
}

export interface UpdateSubscriptionCommand {
  id?: string;
  packageId?: string;
  billingCycle?: BillingCycle;
  autoRenew?: boolean;
}

export interface CancelSubscriptionCommand {
  id: string;
  reason: string;
}

export interface RenewSubscriptionCommand {
  id: string;
  newPackageId?: string;
}

export interface ChangePackageCommand {
  newPackageId: string;
}

class SubscriptionService {
  private readonly basePath = '/api/master/subscriptions';

  /**
   * Get all subscriptions with filtering
   */
  async getAll(query?: GetSubscriptionsQuery): Promise<SubscriptionDto[]> {
    try {
      const params = new URLSearchParams();

      if (query) {
        if (query.tenantId) params.append('tenantId', query.tenantId);
        if (query.status !== undefined) params.append('status', query.status.toString());
        if (query.autoRenew !== undefined) params.append('autoRenew', query.autoRenew.toString());
      }

      const queryString = params.toString();
      const url = queryString ? `${this.basePath}?${queryString}` : this.basePath;

      return await apiClient.get<SubscriptionDto[]>(url);
    } catch (error) {
      console.error('Failed to fetch subscriptions:', error);
      throw error;
    }
  }

  /**
   * Get subscription by ID
   */
  async getById(id: string): Promise<SubscriptionDto | null> {
    try {
      return await apiClient.get<SubscriptionDto>(`${this.basePath}/${id}`);
    } catch (error) {
      console.error(`Failed to fetch subscription ${id}:`, error);
      return null;
    }
  }

  /**
   * Create a new subscription
   */
  async create(command: CreateSubscriptionCommand): Promise<SubscriptionDto> {
    try {
      return await apiClient.post<SubscriptionDto>(this.basePath, command);
    } catch (error) {
      console.error('Failed to create subscription:', error);
      throw error;
    }
  }

  /**
   * Update subscription
   */
  async update(id: string, command: UpdateSubscriptionCommand): Promise<boolean> {
    try {
      const response = await apiClient.put<boolean>(`${this.basePath}/${id}`, command);
      return response === true;
    } catch (error) {
      console.error(`Failed to update subscription ${id}:`, error);
      throw error;
    }
  }

  /**
   * Cancel a subscription
   */
  async cancel(id: string, command: CancelSubscriptionCommand): Promise<boolean> {
    try {
      const response = await apiClient.post<boolean>(`${this.basePath}/${id}/cancel`, command);
      return response === true;
    } catch (error) {
      console.error(`Failed to cancel subscription ${id}:`, error);
      throw error;
    }
  }

  /**
   * Renew a subscription
   */
  async renew(id: string, command: RenewSubscriptionCommand): Promise<boolean> {
    try {
      const response = await apiClient.post<boolean>(`${this.basePath}/${id}/renew`, command);
      return response === true;
    } catch (error) {
      console.error(`Failed to renew subscription ${id}:`, error);
      throw error;
    }
  }

  /**
   * Change subscription package for a tenant
   */
  async changePackage(tenantId: string, command: ChangePackageCommand): Promise<boolean> {
    try {
      const response = await apiClient.post<boolean>(
        `${this.basePath}/tenant/${tenantId}/change-package`,
        command
      );
      return response === true;
    } catch (error) {
      console.error(`Failed to change package for tenant ${tenantId}:`, error);
      throw error;
    }
  }

  /**
   * Helper: Get status color for UI
   */
  getStatusColor(status: SubscriptionStatus): string {
    switch (status) {
      case SubscriptionStatus.Deneme:
        return 'processing';
      case SubscriptionStatus.Aktif:
        return 'success';
      case SubscriptionStatus.Askida:
        return 'warning';
      case SubscriptionStatus.Iptal:
        return 'default';
      case SubscriptionStatus.Suresi_Dolmus:
        return 'error';
      default:
        return 'default';
    }
  }

  /**
   * Helper: Get status display text
   */
  getStatusText(status: SubscriptionStatus): string {
    switch (status) {
      case SubscriptionStatus.Deneme:
        return 'Deneme';
      case SubscriptionStatus.Aktif:
        return 'Aktif';
      case SubscriptionStatus.Askida:
        return 'Askıda';
      case SubscriptionStatus.Iptal:
        return 'İptal';
      case SubscriptionStatus.Suresi_Dolmus:
        return 'Süresi Dolmuş';
      default:
        return 'Bilinmiyor';
    }
  }

  /**
   * Helper: Get billing cycle display text
   */
  getBillingCycleText(cycle: BillingCycle): string {
    switch (cycle) {
      case BillingCycle.Aylik:
        return 'Aylık';
      case BillingCycle.Yillik:
        return 'Yıllık';
      case BillingCycle.Tek_Sefer:
        return 'Tek Sefer';
      default:
        return 'Bilinmiyor';
    }
  }
}

// Export singleton instance
export const subscriptionService = new SubscriptionService();
